src/EFCore.PG/Internal/NpgsqlSingletonOptions.cs

```diff
@@ -76,4 +89,10 @@ public virtual void Validate(IDbContextOptions options)
                     nameof(DbContextOptionsBuilder.UseInternalServiceProvider)));
         }
     }
+
+    // First, try to get a configured data source from the NpgsqlOptionsExtension (configured explicitly via UseNpgsql).
+    // Otherwise check the application service provider to see if one is registered there.
+    static DbDataSource? GetDataSource(IDbContextOptions options, NpgsqlOptionsExtension npgsqlOptions)
+        => npgsqlOptions.DataSource
+            ?? options.FindExtension<CoreOptionsExtension>()?.ApplicationServiceProvider?.GetService<DbDataSource>();
```

Comment:
Safer to do the following
```suggestion
            ?? options.FindExtension<CoreOptionsExtension>()?.ApplicationServiceProvider?.GetService<NpgsqlDataSource>();
```

Reply:
I think it's better to leave the wrapping DbDataSource scenario to manual config. The reason I gave the suggestion is that you don't want to blow up down the line because DbDataSource is actually of type SqliteDataSource and it doesn't quite work the same.

Reply:
But wouldn't that happen only if you don't otherwise explicitly configure EF (i.e. with a connection string), meaning that it will blow up anyway?

Reply:
Discussed offline, we'll start by only resolving NpgsqlDataSource from DI, to minimize confusion in case an incompatible data source is present. Users who want wrappers can just use the explicit option with UseNpgsql.